src/naive/datetime/tests.rs

```diff
@@ -367,3 +367,32 @@ fn test_and_utc() {
     assert_eq!(dt_utc.naive_local(), ndt);
     assert_eq!(dt_utc.timezone(), Utc);
 }
+
+#[test]
+fn test_parse_from_iso8601() {
+    let parse = |s| NaiveDateTime::parse_from_iso8601(s).map(|(dt, _)| dt);
+    let datetime = |y, m, d, h, n, s, nano| {
+        NaiveDate::from_ymd_opt(y, m, d).unwrap().and_hms_nano_opt(h, n, s, nano).unwrap()
+    };
+
+    // Taken from ISO 8601
+    assert_eq!(parse("19850412T101530"), Ok(datetime(1985, 4, 12, 10, 15, 30, 0)));
+    assert_eq!(parse("1985-04-12T10:15:30"), Ok(datetime(1985, 4, 12, 10, 15, 30, 0)));
+    assert_eq!(parse("19850412T1015"), Ok(datetime(1985, 4, 12, 10, 15, 0, 0)));
+    assert_eq!(parse("1985-04-12T10:15"), Ok(datetime(1985, 4, 12, 10, 15, 0, 0)));
+    assert_eq!(parse("1985102T1015"), Ok(datetime(1985, 4, 12, 10, 15, 0, 0)));
+    assert_eq!(parse("1985-102T10:15"), Ok(datetime(1985, 4, 12, 10, 15, 0, 0)));
+    assert_eq!(parse("1985W155T1015"), Ok(datetime(1985, 4, 12, 10, 15, 0, 0)));
+    assert_eq!(parse("1985-W15-5T10:15"), Ok(datetime(1985, 4, 12, 10, 15, 0, 0)));
+    // Test 24:00:00 wraps to the next day
+    assert_eq!(parse("2023-06-09T24:00:00"), Ok(datetime(2023, 6, 10, 0, 0, 0, 0)));
+}
```

Comment:
Needs error variations

    assert_eq!(parse("1985-04-12T10:15:3"), Err(INVALID)); // incomplete seconds
    assert_eq!(parse("1985-04-12T10:1:30"), Err(INVALID)); // incomplete minutes
    assert_eq!(parse("1985-04-12T0:15:30"), Err(INVALID)); // incomplete hours
    assert_eq!(parse("1985-04-1T10:15:30"), Err(INVALID)); // incomplete day
    assert_eq!(parse("1985-4-12T10:15:30"), Err(INVALID)); // incomplete month
    assert_eq!(parse("1985-0-12T10:15:30"), Err(INVALID)); // bad month
    assert_eq!(parse("99999-04-12T10:15:30"), Err(INVALID)); // bad year
    assert_eq!(parse("1985-04-12_10:15:30"), Err(INVALID)); // bad separator
    assert_eq!(parse("1985-04-12💀10:15:30"), Err(INVALID)); // bad separator


I'm not sure which of these are valid `INVALID` asserts, but you get the idea.